Torantulino/Auto-GPT
# Introducing Summarize Steps Mode: Simplified Session Logs for Autonomous Runs and Easy Progress Monitoring

### Background

"Summarize Steps mode" creates brief GPT-generated summaries for each step, and then combines them into a final summary for each run. 

Turning it on allows for concise step summaries to be displayed in the console and saved in log files.

Upon exit, a final summary is generated by creating an agent that takes all the input inside the log and summarizes it. The summary is split into chunks before, so that it doesn't exceed the token limit. 

This feature aids in tracking the assistant's responses and identifying deviations, increasing the potential for improving it's functionality. 

Activate it by running main.py with -ss or --summarize-steps:

```
python scripts/main.py --summarize-steps
```


### Changes

🔀 merge: Merge branch 'feature/summarize-steps-mode' into main
- Added support for new summarize steps mode
- Changes made to five files: ai_config.py, config.py, data.py, json_parser.py and main.py
- Toggle on or off with '-ss' or '--summarize-steps' args when running main.py
- Prompt and JSON schema summarized using step_summarizer module when enabled
- Added final summarization, which condenses all summarized steps into an even shorter summary (via a GPT agent) upon exit

🔧 chore(main.py): refactor interaction loop and add support for step summarization
- Refactored interaction loop to include step summarization support
- Added `get_step_summary` function to extract step summary from assistant's response
- Initialized `Summary` object at beginning of loop if `summarize_steps` argument is passed
- Step summary printed to console and appended to summary file if summarization is enabled
- Final summary appended to summary file upon program exit

✨ feat(step_summarizer.py): add Summary and SummaryUtils classes, implement text chunk splitting, and add tests
- Added Summary and SummaryUtils classes to step_summarizer.py
- Created summary object for printing step summaries to console and appending to file
- SummaryUtils class includes utility methods for splitting sentences into bullet points and updating JSON schema and prompt
- Implemented `append_final_summary` and `split_text_into_chunks` methods for handling large text by splitting it into smaller chunks

### Test Plan

1. Run the test suite with python -m unittest step_summarizer_tests.py to ensure all tests pass.
2. Manually test the main.py script with the -ss or --summarize-steps argument to verify the proper functioning of Summarize Steps mode.
3. Run python scripts/main.py --summarize-steps and observe the output to ensure step summaries are printed to the console
4. Check the generated summary file to verify that step summaries and the final summary are correctly appended.
5. Test the script without the -ss or --summarize-steps argument to ensure normal functionality is preserved when the new mode is not enabled
6. Test various input texts and scenarios to confirm that the summaries are coherent and useful for different use cases.

### Change Safety

- [✓] I have added tests to cover my changes
- [✓] I have considered potential risks and mitigations for my changes

### Potential risks and mitigations:

Risk: Summaries may not always be coherent or helpful due to GPT's inherent limitations.
Mitigation: Carefully monitor the output to ensure the feature is working as expected and adjust the prompts and code as needed to improve the quality of summaries.

Risk: Large texts or numerous steps may cause the script to take a long time to generate summaries or hit API rate limits.
Mitigation: Implement chunking to handle large texts and limit the number of steps during testing. Additionally, monitor API usage to avoid unexpected costs or throttling.

Risk: Unexpected bugs or issues may arise when the new mode is enabled or when working with the modified code.
Mitigation: Thoroughly test the changes and monitor the output during usage. Address any issues as they arise and consider adding more tests to cover edge cases.
